ktor-utils/jvm/src/io/ktor/util/reflect/ServiceLoader.kt

```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright 2014-2024 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.util.reflect
+
+import io.ktor.utils.io.*
+import java.util.*
+
+/**
+ * Loads implementations of service [T] using [ServiceLoader.load].
+ *
+ * NOTE: ServiceLoader should use specific call convention to be optimized by R8 on Android:
+ * `ServiceLoader.load(X.class, X.class.getClassLoader()).iterator()`
+ * See: https://r8.googlesource.com/r8/+/refs/heads/main/src/main/java/com/android/tools/r8/ir/optimize/ServiceLoaderRewriter.java
+ */
+@InternalAPI
+public inline fun <reified T> loadService(): Iterable<T> = Iterable {
+    ServiceLoader.load(
+        T::class.java,
+        T::class.java.classLoader,
+    ).iterator()
+}
```

Comment:
I've tried to replace it with the following code, to retrieve iterator only once:
```kotlin
public inline fun <reified T> loadService(): Iterable<T>
    val iterator = ServiceLoader.load(
        T::class.java,
        T::class.java.classLoader,
    ).iterator()
    
    return Iterable { iterator }
}
```

But for some reason this change [breaks serialization tests](https://ktor.teamcity.com/buildConfiguration/Ktor_KtorMatrixCoreLinuxJava21/291227?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&expandBuildTestsSection=true&hideProblemsFromDependencies=false&expandPull+Request+Details=true&expandBuildChangesSection=true) 🤔

Reply:
iterators are single use, so they can not be used like this and looks like in serialization tests,  result of `loadService()` is called multiple times and so when it's called second time, it will immediately start returning `hasNext=false`. (e.g https://pl.kotl.in/JkZSEveSX)

So looks like it's better to use `Iterable { ... }.toList()` as it was before

Reply:
Ah, it explains everything. I'll add two separate functions then: one to load all services and another to load only the first.

Reply:
Fixed 👉  https://github.com/ktorio/ktor/pull/4454/commits/1eeba0739dab72ce5b3d0f23ebec351656acfbd3

Reply:
LGTM ^_^